internal/constants/scripts.go

```diff
@@ -1,58 +1,163 @@
-    host_macs+=($mac)
+    host_macs_to_iface[$mac]=$nic
   done
 }
 
-function copy_connection_files_by_mac_address() {
-  # iterate over nmconnection files to find associated file with current host
-  for nmconn_file in $(ls -1 ${NMCONNECTIONS_DIR}/*.nmconnection)
+function find_host_directory_by_mac_address() {
+  for d in $(ls -d ${NMCONNECTIONS_DIR}/host*)
   do
-    # get mac-address from connection files to search by
-    mac_address=$(grep -A1 '\[802-3-ethernet\].*' $nmconn_file | grep -v "802-3-ethernet" | cut -d= -f2 | tr '[:lower:]' '[:upper:]')
-    echo "mac adddress of ${nmconn_file} is ${mac_address}"
-
-    # compare mac-address from nmconnection with host's mac addresses
-    if [[ " ${host_macs[@]} " =~ " ${mac_address} " ]]; then
-      echo "copying ${nmconn_file} to NM working directory ${ETC_NETWORK_MANAGER}"
-      cp $nmconn_file ${ETC_NETWORK_MANAGER}/
+    mapping_file="${d}/${MAC_NIC_MAPPING_FILE}"
+    if [ ! -f $mapping_file ]
+    then
+      echo "Mapping file '$mapping_file' is missing. Skipping on directory '$d'" | systemd-cat
+      continue
     fi
+
+    # check if mapping file contains mac-address that exists on the current host
+    for mac_address in $(cat $mapping_file | cut -d= -f1 | tr '[:lower:]' '[:upper:]')
+    do
+      if [[ ! -z "${host_macs_to_iface[${mac_address}]:-}" ]]
+      then
+        host_dir=$d
+        return
+      fi
+    done
   done
+
+  if [ -z "$host_dir" ]
+  then
+    echo "None of host direcotries are a match for the current host" | systemd-cat
+    exit 0
+  fi
+}
+
+function set_nic_mac_mapping() {
+  # initialize nic_mac_map with mapping file entries
+  readarray -t lines < "${mapping_file}"
+  for line in "${lines[@]}"
+  do
+    mac=${line%%=*}
+    nic=${line#*=}
+    nic_mac_map[$nic]=${mac^^}
+  done
+}
+
+# Replace logical interface name in nmconnection files with the interface name from the mapping file
+# of host's directory. Replacement is done based on mac-address matching
+function update_interface_names_by_mapping_file() {
+
+  # iterate over host's nmconnection files and replace logical interface name with host's nic name
+  for nmconn_file in $(ls -1 ${host_dir}/*.nmconnection)
+  do
+    # iterate over mapping to find nmconnection files with logical interface name
+    for nic in "${!nic_mac_map[@]}"
+    do
+      mac=${nic_mac_map[$nic]}
+
+      # the pattern should match '=eth0' (interface name) or '=eth0.' (for vlan devices)
+      if grep -q -e "=$nic$" -e "=$nic\." "$nmconn_file"
+      then
+        # get host interface name
+        host_iface=${host_macs_to_iface[$mac]}
+        if [ -z "$host_iface" ]
+        then
+          echo "Mapping file contains MAC Address '$mac' (for logical interface name '$nic') that doesn't exist on the host" | systemd-cat
+          continue
+        fi
+
+        # replace logical interface name with host interface name
+        sed -i -e "s/=$nic$/=$host_iface/g" -e "s/=$nic\./=$host_iface\./g" $nmconn_file
```

Comment:
maybe we store the result in the new temp file and not the original one? that way it will allow us a better debugging, at least in the earlier stages. what do you think?

Reply:
I'll change the script to create a copy of the entire directory relevant for the host (identified by `host_dir`).
The original files will remain intact, and the rest of the script will not have to be changed too much